Source/WebCore/html/HTMLAttachmentElement.cpp

```diff
@@ -146,28 +175,72 @@ void HTMLAttachmentElement::ensureModernShadowTree(ShadowRoot& root)
+
+    updateSave(attributeWithoutSynchronization(saveQualifiedName()));
+}
 
-    m_elementWithAction = HTMLDivElement::create(document());
-    m_elementWithAction->setIdAttribute(attachmentActionIdentifier());
-    if (const auto& action = attachmentActionForDisplay(); !action.isEmpty())
-        m_elementWithAction->setInnerText(String { action });
-    container->appendChild(*m_elementWithAction);
+class AttachmentSaveEventListener final : public EventListener {
+public:
+    static Ref<AttachmentSaveEventListener> create(HTMLAttachmentElement& attachment) { return adoptRef(*new AttachmentSaveEventListener(attachment)); }
 
-    m_elementWithTitle = HTMLDivElement::create(document());
-    m_elementWithTitle->setIdAttribute(attachmentTitleIdentifier());
-    if (auto title = attachmentTitleForDisplay(); !title.isEmpty())
-        m_elementWithTitle->setInnerText(WTFMove(title));
-    container->appendChild(*m_elementWithTitle);
+    bool operator==(const EventListener& other) const final
+    {
+        return this == &other;
+    }
 
-    m_elementWithSubtitle = HTMLDivElement::create(document());
-    m_elementWithSubtitle->setIdAttribute(attachmentSubtitleIdentifier());
-    if (auto subtitle = attachmentSubtitleForDisplay(); !subtitle.isEmpty())
-        m_elementWithSubtitle->setInnerText(WTFMove(subtitle));
-    container->appendChild(*m_elementWithSubtitle);
+    void handleEvent(ScriptExecutionContext&, Event& event) final
+    {
+        if (event.type() == eventNames().clickEvent) {
+            event.preventDefault();
+            event.stopPropagation();
+            event.stopImmediatePropagation();
+
+            CustomEvent::Init init;
+            init.bubbles = true;
+            init.cancelable = true;
+            init.composed = true;
+            m_attachment->dispatchEvent(CustomEvent::create(m_attachment->attributeWithoutSynchronization(saveQualifiedName()), init, event.isTrusted() ? Event::IsTrusted::Yes : Event::IsTrusted::No));
+        } else
+            ASSERT_NOT_REACHED();
+    }
+
+private:
+    explicit AttachmentSaveEventListener(HTMLAttachmentElement& attachment)
+        : EventListener(CPPEventListenerType)
+        , m_attachment(attachment)
+    {
+    }
+
+    WeakPtr<HTMLAttachmentElement, WeakPtrImplWithEventTargetData> m_attachment;
+};
+
+void HTMLAttachmentElement::updateSave(const AtomString& eventTypeName)
+{
+    if (!m_elementContainer)
+        return;
+
+    if (eventTypeName.isEmpty()) {
+        if (m_elementWithSave) {
+            m_elementContainer->removeChild(*m_elementWithSave);
+            m_elementWithSave = nullptr;
+        }
+        return;
+    }
+
+    if (!m_elementWithSave) {
+        m_elementWithSave = createShadowDivElement(attachmentSaveIdentifier(), String { });
+        m_elementWithSave->addEventListener(eventNames().clickEvent, AttachmentSaveEventListener::create(*this), { });
+    }
```

Comment:
```suggestion
    if (!m_elementContainer)
        return;
        
    if (m_elementWithSave) {
        if (eventTypeName.isEmpty()) {
            m_elementContainer->removeChild(*m_elementWithSave);
            m_elementWithSave = nullptr;
        }
        return;
    }

    m_elementWithSave = createShadowDivElement(attachmentSaveIdentifier(), String { });
    m_elementWithSave->addEventListener(eventNames().clickEvent, AttachmentSaveEventListener::create(*this), { });
```

Reply:
Your suggested code doesn't correctly handle the case where `m_elementWithSave` is null and the `eventTypeName` is empty: It would go to line 237 and always create the save button, when it shouldn't (because of the empty `eventTypeName`). We'd need to add `if (!eventTypeName.isEmpty())` around lines 237-238, which goes back to the double-if that I think you were trying to avoid.

Reply:
Ah I guess your code is fine then. Btw, we should probably check against a null string rather than an empty string if possible.